File: leetcode_75/Monotonic_Stack/cpp/739.cpp
#include <vector>
#include <stack>

using namespace std;

class Solution
{
public:
    vector<int> dailyTemperatures(vector<int> &temperatures)
    {
        int n = temperatures.size();
        vector<int> result(n, 0);

        stack<pair<int, int>> st;
        st.push({temperatures[0], 0});
        for (int i = 1; i < n; ++i)
        {
            while (!st.empty() && st.top().first < temperatures[i])
            {
                result[st.top().second] = i - st.top().second;
                st.pop();
            }

            st.push({temperatures[i], i});
        }

        return result;
    }
};

File: leetcode_75/Monotonic_Stack/python/739.py
from typing import List


class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        n = len(temperatures)
        result = [0] * n

        stack = [(temperatures[0], 0)]
        for i in range(1, n):
            while stack and stack[-1][0] < temperatures[i]:
                data = stack.pop()
                result[data[1]] = (i - data[1])

            stack.append((temperatures[i], i))

        return result
